File: Client Frontend/client_frontend/credentials/oauth_pkce.js
import { createHash, randomBytes } from "node:crypto";

function toBase64Url(buffer) {
  return buffer.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

export function generateCodeVerifier(length = 64) {
  const targetLength = Math.max(43, Number(length) || 64);
  return toBase64Url(randomBytes(targetLength)).slice(0, targetLength);
}

export function buildCodeChallenge(verifier) {
  return toBase64Url(createHash("sha256").update(String(verifier || ""), "ascii").digest());
}

export function generateState() {
  return toBase64Url(randomBytes(24));
}

File: 00 - Orchestrator/orchestrator/credentials/oauth_pkce.py
"""PKCE helpers for browser-based Orchestrator OAuth login."""

from __future__ import annotations

import base64
import hashlib
import secrets


def generate_code_verifier(length: int = 64) -> str:
    if length < 43:
        raise ValueError("PKCE verifier length must be >= 43")
    raw = secrets.token_urlsafe(length)
    return raw[: max(length, 43)]


def build_code_challenge(verifier: str) -> str:
    digest = hashlib.sha256(verifier.encode("ascii")).digest()
    return base64.urlsafe_b64encode(digest).rstrip(b"=").decode("ascii")


def generate_state() -> str:
    return secrets.token_urlsafe(24)
